support-lib/jni/DataRef_jni.cpp

```diff
@@ -79,12 +79,15 @@ class DataRefJNI : public DataRef::Impl {
     // wrap a ByteBuffer object from java
     explicit DataRefJNI(jobject data) {
         auto* env = jniGetThreadEnv();
+        auto len = env->GetDirectBufferCapacity(data);
+        if (len == -1) {
+            // GetDirectBufferCapacity() returns -1 when the ByteBuffer is not direct
+            throw std::invalid_argument("ByteBuffer is not allocated with allocateDirect()");
+        }
         _data = {env, data};
         // call ByteBuffer.isReadOnly() to determine mutability
         _readonly = env->CallBooleanMethod(_data.get(), JniClass<BufferClassInfo>::get().isReadOnly) != 0;
         jniExceptionCheck(env);
-        auto len = env->GetDirectBufferCapacity(_data.get());
-        assert(len != -1); // GetDirectBufferCapacity() returns -1 when the ByteBuffer is not direct
```

Comment:
You could also keep the assert in, above the throw statement. That way for NDEBUG builds, you'll hit the exception rather than crash, and for DEBUG builds, you'll get the assert. 

Also we should make sure all asserts have a message, `assert(len != 1 && "DataRef must be used with a direct-allocated ByteBuffer from java")`

Reply:
Seems redundant?  What is something assert provides but exception does not?

Reply:
This is native code in Java so we are not going to see the debugger stopping at the assertion line anyway.

Reply:
Obviously there's trade offs between them. I was just assuming that for DEBUG builds the assert was preferable over the exception